cpp/src/copying/copy.cu

```diff
@@ -183,55 +159,89 @@ class logical_not {
+                                               is_left);
+
+  auto const scatter_src_lhs = cudf::detail::gather(table_view{std::vector<column_view>{lhs}},
+                                                    scatter_map.begin(),
+                                                    scatter_map_end,
+                                                    out_of_bounds_policy::DONT_CHECK,
+                                                    stream);
+
+  auto result = cudf::detail::scatter(
+    table_view{std::vector<column_view>{scatter_src_lhs->get_column(0).view()}},
+    scatter_map.begin(),
+    scatter_map_end,
+    table_view{std::vector<column_view>{rhs}},
+    false,
+    stream,
+    mr);
+
+  return std::move(result->release()[0]);
+}
 
-  // Bail out for Scalars.
-  // For nested types types, scatter/gather based copy_if_else() is not currently supported
-  // if either `lhs` or `rhs` is a scalar, partially because:
-  //   1. Struct scalars are not yet available.
-  //   2. List scalars do not yet support explosion to a full column.
-  CUDF_FAIL("Scalars of nested types are not currently supported!");
-  (void)lhs;
-  (void)rhs;
-  (void)size;
-  (void)is_left;
-  (void)stream;
-  (void)mr;
+template <typename Filter>
+std::unique_ptr<column> scatter_gather_based_if_else(cudf::scalar const& lhs,
+                                                     cudf::column_view const& rhs,
+                                                     size_type size,
+                                                     Filter is_left,
+                                                     rmm::cuda_stream_view stream,
+                                                     rmm::mr::device_memory_resource* mr)
+{
+  auto scatter_map = rmm::device_uvector<size_type>{static_cast<std::size_t>(size), stream};
+  auto const scatter_map_end = thrust::copy_if(rmm::exec_policy(stream),
+                                               thrust::make_counting_iterator(size_type{0}),
+                                               thrust::make_counting_iterator(size_type{size}),
+                                               scatter_map.begin(),
+                                               is_left);
+
+  auto const scatter_map_size  = scatter_map_end - scatter_map.begin();
+  auto scatter_source          = std::vector<std::reference_wrapper<const scalar>>{std::ref(lhs)};
+  auto scatter_map_column_view = cudf::column_view{cudf::data_type{cudf::type_id::INT32},
+                                                   static_cast<cudf::size_type>(scatter_map_size),
+                                                   scatter_map.begin()};
+
+  auto result = cudf::scatter(
+    scatter_source, scatter_map_column_view, table_view{std::vector<column_view>{rhs}}, false, mr);
+
+  return std::move(result->release()[0]);
+}
+
+template <typename Filter>
+std::unique_ptr<column> scatter_gather_based_if_else(cudf::column_view const& lhs,
```

Comment:
It breaks my heart that these overloads can't be collapsed into a single function. I'll play with this a little bit and see if I can't come up with something.

Reply:
I think I see the problem. The `detail::scatter()` APIs are not symmetric across the `table_view` and `vector<reference<scalar>>` overloads. :/
If we had an overload of `detail::scatter(vector<reference<scalar>>, ...)` that took `[scatter_map_begin, scatter_map_end)` as arguments instead of a `column_view`, we stood a chance of collapsing these 4 overloads into one or two. 

I think it can be done, based on the definitions I see in [scatter.cu](https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/src/copying/scatter.cu#L308).

That said, I'm not opposed to leaving this the way it is, if it looks like too much change.

Reply:
My preference is to do this refactoring as a separate PR